### PR TITLE
MDIO v1 Templates and Template Registry

### PR DESCRIPTION
**Pull request by dmitriyrepin**

### CODE REVIEW

File: src/mdio/schemas/v1/templates/abstract_dataset_template.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+    def _add_dimensions(self) -> None:
+        """A virtual method that can be overwritten by subclasses to add custom dimensions.
+
+        Uses the class field 'builder' to add dimensions to the dataset.
+        """
+        for i in range(len(self._dim_names)):
+            self._builder.add_dimension(self._dim_names[i], self._dim_sizes[i])
+
+    def _add_coordinates(self) -> None:
```
Comment (tasansal) [inline src/mdio/schemas/v1/templates/abstract_dataset_template.py:121]: It is not immediately obvious what the default behavior of this method is. Should we make it into an abstract method and enforce people to add the coordinates? It would also clean up the `AbstractDatasetTemplate.__init__`. Thoughts?

File: docs/template_registry.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    print(f"Registered template named {template_name}")
+
+    print(f"Registered templates: {list_templates()}")
+
+# Application startup
+setup_standard_templates()
+
+# Later in the application
+template = TemplateRegistry().get_template("PostStack3DDepth")
+dataset = template.create_dataset(name="Campos 3d m/m/ft",
```
Comment (tasansal) [inline docs/template_registry.md:255]: please remove campos references from all files

File: tests/unit/v1/templates/test_seismic_3d_poststack.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        assert t2._trace_domain == "elevation"
+        assert t2.get_name() == "PostStack3DElevation"
+
+    def test_build_dataset_depth(self) -> None:
+        """Unit tests for Seismic3DPostStackTemplate build with depth domain."""
+        t = Seismic3DPostStackTemplate(domain="depth")
+
+        assert t.get_name() == "PostStack3DDepth"
+        dataset = t.build_dataset(
+            "Campos 3D", sizes=[256, 512, 1024], coord_units=[_UNIT_METER, _UNIT_METER]
```
Comment (tasansal) [inline tests/unit/v1/templates/test_seismic_3d_poststack.py:88]: remove campos reference

File: tests/unit/v1/templates/test_seismic_3d_poststack.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        assert seismic.metadata.chunk_grid.configuration.chunk_shape == [128, 128, 128]
+        assert seismic.metadata.stats_v1 is None
+
+    def test_build_dataset_time(self) -> None:
+        """Unit tests for Seismic3DPostStackTimeTemplate build with time domain."""
+        t = Seismic3DPostStackTemplate(domain="time")
+
+        assert t.get_name() == "PostStack3DTime"
+        dataset = t.build_dataset(
+            "Campos 3D", sizes=[256, 512, 1024], coord_units=[_UNIT_METER, _UNIT_METER]
```
Comment (tasansal) [inline tests/unit/v1/templates/test_seismic_3d_poststack.py:139]: remove campos reference

File: .devcontainer/Dockerfile.dev
Comment (tasansal) [inline .devcontainer/Dockerfile.dev:1]: please keep devcontainer related changes in a separate PR

File: .devcontainer/devcontainer.json
Comment (tasansal) [inline .devcontainer/devcontainer.json:1]: please keep devcontainer related changes in a separate PR

File: docs/template_registry.md
Comment (tasansal) [inline docs/template_registry.md:1]: this also need to be added to the documentation index to render as a page. currently this won't get rendered in docs.

File: .devcontainer/Dockerfile.dev
Comment (dmitriyrepin) [reply .devcontainer/Dockerfile.dev:1]: Reverted. Sorry for the leak of the local dev environment.

File: .devcontainer/devcontainer.json
Comment (dmitriyrepin) [reply .devcontainer/devcontainer.json:1]: Reverted. Sorry for the leak of the local dev environment.

File: docs/template_registry.md (hunk 2 above)
Comment (dmitriyrepin) [reply docs/template_registry.md:255]: Replaced with Seismic

File: tests/unit/v1/templates/test_seismic_3d_poststack.py (hunk 3 above)
Comment (dmitriyrepin) [reply tests/unit/v1/templates/test_seismic_3d_poststack.py:88]: Replaced

File: tests/unit/v1/templates/test_seismic_3d_poststack.py (hunk 4 above)
Comment (dmitriyrepin) [reply tests/unit/v1/templates/test_seismic_3d_poststack.py:139]: Replaced

File: src/mdio/schemas/v1/templates/template_registry.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    def __init__(self) -> None:
+        if not self._initialized:
+            with self._lock:
+                if not self._initialized:
+                    self._templates: dict[str, AbstractDatasetTemplate] = {}
+                    self._registry_lock = threading.RLock()
+                    TemplateRegistry._initialized = True
+
+    def register(self, instance: AbstractDatasetTemplate) -> str:
```
[the remark this answers is not in the document]
Comment (dmitriyrepin) [reply src/mdio/schemas/v1/templates/template_registry.py:33]: The standard approach is to have two arguments: the name under which template will be registered and template instance:
```Python
  def register(self, name: str, instance: AbstractDatasetTemplate) -> None
```
This allows for the following 'tricks':
```Python
registry.resfister("PostStack2DDepth", Seismic3DPostStackTemplate("time"))
registry.resfister("PostStack3DTime", Seismic2DPostStackTemplate("depth"))
```
My approach forces the proper registration:
```Python
name1 = registry.resfister(Seismic3DPostStackTemplate("time")) # "PostStack3DTime"
name2 = registry.resfister(Seismic2DPostStackTemplate("depth")) # "PostStack2DDepth"
```

As for the return value:  
You might need the name at the point of registration. With my approach, you do not know it for sure until the template is instantiated. If you need it, returning the name shortens the code. Otherwise you would have to do the following:

Returning template
```Python
t1 = registry.resfister(Seismic3DPostStackTemplate("time"))
name1 = t1.name
t2 = registry.resfister(Seismic2DPostStackTemplate("depth")) 
name2  = t2.name
```
or returning None
```Python
t1 = Seismic3DPostStackTemplate("time")
registry.resfister(t1)
name1 = t1.name
t2 = Seismic2DPostStackTemplate("depth")
registry.resfister(t2) # "PostStack2DDepth"
name2  = t2.name
```

However, I am open to changing the approach.

File: docs/template_registry.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+Alternative way to get the singleton instance.
+
+- **Returns:** The singleton registry instance
+
+#### `reset_instance() -> None`
+
+Resets the singleton instance. Useful for testing to ensure clean state.
+
+**Warning:** This should only be used in tests, not in production code.
```
[the remark this answers is not in the document]
Comment (dmitriyrepin) [reply docs/template_registry.md:147]: I removed the whole ##Testing chapter from the docs. 
The template_registry users do not need to know implementation details on how we test it.

File: src/mdio/schemas/v1/templates/template_registry.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        """Get a registered template instance."""
+        with self._registry_lock:
+            name = template_name.lower()
+            if name not in self._templates:
+                err = f"Template '{name}' is not registered."
+                raise KeyError(err)
+            return self._templates[name]
+
+    def unregister(self, template_name: str) -> None:
+        """Unregister a template instance."""
```
[the remark this answers is not in the document]
Comment (dmitriyrepin) [reply src/mdio/schemas/v1/templates/template_registry.py:57]: Corrected

File: src/mdio/schemas/v1/templates/template_registry.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        with self._registry_lock:
+            name = instance.get_name().lower()
+            if name in self._templates:
+                err = f"Template '{name}' is already registered."
+                raise ValueError(err)
+            self._templates[name] = instance
+        return name
+
+    def get(self, template_name: str) -> AbstractDatasetTemplate:
+        """Get a registered template instance."""
```
[the remark this answers is not in the document]
Comment (dmitriyrepin) [reply src/mdio/schemas/v1/templates/template_registry.py:48]: Added 'Arguments',  'Raises' and 'Returns' to all docstrings

File: docs/template_registry.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+The `TemplateRegistry` implements the singleton pattern to ensure there's only one instance managing all dataset templates throughout the application lifecycle. This provides a centralized registry for template management with thread-safe operations.
+
+## Features
+
+- **Singleton Pattern**: Ensures only one registry instance exists
+- **Thread Safety**: All operations are thread-safe using locks
+- **Case Insensitive**: Template names are normalized to lowercase
+- **Global Access**: Convenient global functions for common operations
+- **Testing Support**: Reset functionality for clean test isolation
```
[the remark this answers is not in the document]
Comment (dmitriyrepin) [reply docs/template_registry.md:15]: Doc updated

File: docs/template_registry.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
+# Template Registry Singleton
+
+A thread-safe singleton registry for managing dataset templates in MDIO applications.
```
[the remark this answers is not in the document]
Comment (dmitriyrepin) [reply docs/template_registry.md:3]: As you can tell from the extensive documentation, I have had help from AI while creating the registry.
The muti-threading support was added by AI and I decided to keep it, but we can as well remove it now and add it later, if ever needed.

[the remark this answers is not in the document]
Comment (BrianMichell) [reply docs/template_registry.md:3]: I think it's fine as-is and can denote that this _may not_ be safe in a multi-processing context.

File: src/mdio/schemas/v1/templates/template_registry.py
[the remark this answers is not in the document]
Comment (BrianMichell) [reply src/mdio/schemas/v1/templates/template_registry.py:1]: My preference would be automatic registration and have the end-user clear or override the existing. This helps enforce a standard form of exchange with low friction and fewer lines of boilerplate code.

[the remark this answers is not in the document]
Comment (dmitriyrepin) [reply src/mdio/schemas/v1/templates/template_registry.py:1]: Ok. The default templates are initialized in __init__.py

File: docs/template_registry.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (BrianMichell) [reply docs/template_registry.md:3]: I think your understanding is reversed. The GIL is for multi-threading and a new interpreter is started for each process in multi-processing. 

Since our parallel code uses spawn internally (instead of fork) we would lose the state of the Registry object between processes. This is a non-issue for 99% of use-cases I can think of.

File: src/mdio/schemas/v1/templates/template_registry.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BrianMichell) [reply src/mdio/schemas/v1/templates/template_registry.py:44]: I understand your approach now. I am fine with returning `str: name`. @tasansal Resolve this if you agree?

File: src/mdio/schemas/v1/templates/abstract_dataset_template.py (hunk 1 above)
Comment (tasansal) [reply src/mdio/schemas/v1/templates/abstract_dataset_template.py:121]: i think its ok with float64 for now. hopefully these won't be overridden that much!

File: src/mdio/schemas/v1/templates/template_registry.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tasansal) [reply src/mdio/schemas/v1/templates/template_registry.py:44]: name string or None are either fine by me. we can leave it at str: name since its already implemented

**Review by tasansal — CHANGES_REQUESTED** [review]
Hi @dmitriyrepin it looks great now. But as before, I think the tests are too granular and we are adding a lot of test code to maintain. Can we reduce the tests to hit 100% with higher level tests?

**Comment by BrianMichell** [comment]
LGTM. 

I think we can look to add a new task to reduce all test code granularity towards the end of the 1.0.0 release window.

**Comment by dmitriyrepin** [comment]
> Hi @dmitriyrepin it looks great now. But as before, I think the tests are too granular and we are adding a lot of test code to maintain. Can we reduce the tests to hit 100% with higher level tests?

There 2.5 ways we can take to achieve this:
* 1 ) Reduce the number of asserts.
For example, 
    * in test_buid_dataset_time() we can check only the number of the variables created and do not check the structure of the created variables.
    * For templates, we can keep test_build_dataset_depth() and drop test_build_dataset_time(), since they are executing almost identical functionality.

* 2 ) Do not tests some of the functionality we do not expect to use. 
For example, we could 
    * remove tests for parameter validation
    * remove tests for the edge cases, e.g., test_get_nonexistent_template()
    * remove all multi-threaded tests for template_registry.
    
* 2.5) We can combine a few separate unit tests into a single one: 
For example, in test_template_registry.py we can combine into a single test. 
    * test_registry_template
    * test_register_duplicate_template

Please let me know which way you want to proceed.
